torch/testing/_internal/common_utils.py

```diff
@@ -1128,6 +1128,28 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
     def safeToDense(self, t):
         return t.coalesce().to_dense()
 
+    # Compares torch function with reference function for given sample input (object of SampleInput)
+    # Note: only values are compared, type comparison is not done here
+    def compare_with_reference(self, torch_fn, ref_fn, sample_input, **kwargs):
+        # Note: if .numpy() fails for some reason, we won't know if it failed for inp/args/kwargs (TODO)
+        n_inp, n_args, n_kwargs = sample_input.numpy()
+        t_inp, t_args, t_kwargs = sample_input.input, sample_input.args, sample_input.kwargs
+
+        actual = torch_fn(t_inp, *t_args, **t_kwargs)
+        expected = ref_fn(n_inp, *n_args, **n_kwargs)
+
+        if isinstance(actual, torch.Tensor):
+            self.assertEqual(actual, torch.as_tensor(expected, dtype=actual.dtype), exact_device=False)
+        elif is_iterable_of_tensors(actual):
+            msg = "Expected both functions to return outputs of same size but" \
+                  f" got {len(actual)} from torch and {len(expected)} from NumPy."
+            assert len(actual) == len(expected), msg
+            for x, y in zip(actual, expected):
+                self.assertEqual(x, torch.as_tensor(y, dtype=x.dtype), exact_device=False)
```

Comment:
**Why looping?** (`for x, y in zip(actual, expected)`)

- `actual.dtype` will not work if actual is an iterable of tensors/iterator. (For example: `torch.linalg.qr` returns tuple of tensors)
- `actual.device` will also not work (same reason as above)
- Looping is required since we are doing `torch.as_tensor(y, dtype=x.dtype)` for each tensor in the iterator of tensors.

Note: `actual.cpu()` will also not work for iterable of tensors, hence `exact_device=False`. For consistency, other conditions also use `exact_device=False`.